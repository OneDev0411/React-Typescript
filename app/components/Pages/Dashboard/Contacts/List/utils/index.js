import _ from 'underscore'

export function normalizeFilters(filters) {
  const criteria = []

  _.each(filters, filter => {
    _.each(filter.values, value => {
      criteria.push({
        value,
        invert: filter.operator.invert === true,
        attribute_def: filter.id
      })
    })
  })

  return criteria
}
