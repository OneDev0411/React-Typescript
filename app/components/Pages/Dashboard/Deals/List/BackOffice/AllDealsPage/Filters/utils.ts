import { isValid, parseISO } from 'date-fns'

import {
  DEALS_STATUSES,
  QUERY_ARRAY_PARAM_SPLITTER_CHAR
} from '../../constants'
import { DealsListContext, TDealsStatusList } from '../../types'

// Convert status filter to query string
export const stringifyStatusFilter = (filters: TDealsStatusList): string => {
  const statusKeys = Object.keys(DEALS_STATUSES)
  const values = statusKeys.map(key => `${+!!filters[key]}${key}`)

  return values.join(',')
}

// Convert status filter query string to status object
export const parseStatusFilterString = (
  queryParam: string
): TDealsStatusList => {
  const statusKeys = Object.keys(DEALS_STATUSES)
  const statusQueries = queryParam.split(QUERY_ARRAY_PARAM_SPLITTER_CHAR)

  const statusList: TDealsStatusList = {}

  statusQueries.forEach(statusQuery => {
    const key = statusQuery.slice(1)
    const value = !!statusQuery.slice(0, 1)

    if (statusKeys.includes(key)) {
      statusList[key] = value
    }
  })

  return statusList
}

// Convert Range Date filter to query string
export const stringifyRangeDateFilter = (filters: DealsListContext): string => {
  if (filters.date && (filters.date.from || filters.date.to)) {
    return `${filters.date.from ?? ''}${QUERY_ARRAY_PARAM_SPLITTER_CHAR}${
      filters.date.to ?? ''
    }`
  }

  return ''
}

// Convert Range Date filter query string to Range date object
export const parseRangeDateFilterString = (
  queryParam: string
): DealsListContext => {
  const dateQueries = queryParam.split(QUERY_ARRAY_PARAM_SPLITTER_CHAR)

  const from: string | undefined = isValid(parseISO(dateQueries[0]))
    ? (dateQueries[0] as string)
    : undefined

  const to: string | undefined = isValid(parseISO(dateQueries[1]))
    ? (dateQueries[1] as string)
    : undefined

  return { date: { from, to } }
}
