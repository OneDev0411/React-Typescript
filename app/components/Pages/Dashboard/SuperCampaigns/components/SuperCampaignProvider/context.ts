import { createContext } from 'react'

export const SuperCampaignDetailContext =
  createContext<Optional<ISuperCampaign<'template_instance'>>>(undefined)
