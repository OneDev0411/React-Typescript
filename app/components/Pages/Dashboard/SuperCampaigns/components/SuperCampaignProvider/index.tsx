export { default as SuperCampaignProvider } from './SuperCampaignProvider'
export * from './use-super-campaign'
