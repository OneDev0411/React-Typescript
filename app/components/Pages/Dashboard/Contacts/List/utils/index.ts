import _ from 'underscore'

import { FLOW_FILTER_ID, OPEN_HOUSE_FILTER_ID } from 'crm/List/constants'

export const normalizeAttributeFilters = (
  filters: StringMap<IActiveFilter>
): IContactAttributeFilter[] => {
  const criteria: IContactAttributeFilter[] = []

  _.each(filters, filter => {
    if (isAttributeFilter(filter)) {
      _.each(filter.values, ({ value }) => {
        criteria.push({
          value,
          invert: filter.operator.invert === true,
          attribute_def: filter.id
        })
      })
    }
  })

  return criteria
}

export function isAttributeFilter(filter) {
  return ![OPEN_HOUSE_FILTER_ID, FLOW_FILTER_ID].includes(filter.id)
}
