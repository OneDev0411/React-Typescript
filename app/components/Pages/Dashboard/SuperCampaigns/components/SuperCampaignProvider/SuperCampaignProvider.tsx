import { ReactNode } from 'react'

import { SuperCampaignDetailContext } from './context'

interface SuperCampaignDetailProviderProps {
  superCampaign: ISuperCampaign<'template_instance'>
  children: ReactNode
}

function SuperCampaignDetailProvider({
  superCampaign,
  children
}: SuperCampaignDetailProviderProps) {
  return (
    <SuperCampaignDetailContext.Provider value={superCampaign}>
      {children}
    </SuperCampaignDetailContext.Provider>
  )
}

export default SuperCampaignDetailProvider
