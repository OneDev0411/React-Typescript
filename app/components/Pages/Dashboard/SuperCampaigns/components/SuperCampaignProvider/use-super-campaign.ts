import { useContext } from 'react'

import { SuperCampaignDetailContext } from './context'

export function useSuperCampaign() {
  const context = useContext(SuperCampaignDetailContext)

  if (!context) {
    throw new Error(
      'The useSuperCampaignDetail must be used within SuperCampaignDetailProvider'
    )
  }

  return context
}
