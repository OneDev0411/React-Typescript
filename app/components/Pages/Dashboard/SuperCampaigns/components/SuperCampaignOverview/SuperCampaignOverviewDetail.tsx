import { makeStyles } from '@material-ui/core'

import { useReplaceQueryParam } from '@app/hooks/use-query-param'
import { useUpdateSuperCampaign } from '@app/models/super-campaign'

import { isSuperCampaignReadOnly } from '../../helpers'
import SuperCampaignCard from '../SuperCampaignCard'
import SuperCampaignCardHeader from '../SuperCampaignCardHeader'
import SuperCampaignDrawer from '../SuperCampaignDrawer'
import { SuperCampaignFormValues } from '../SuperCampaignDrawer/types'
import { useSuperCampaign } from '../SuperCampaignProvider'
import SuperCampaignTemplate from '../SuperCampaignTemplate/SuperCampaignTemplate'

import SuperCampaignOverviewDetailLabelValue from './SuperCampaignOverviewDetailLabelValue'

const useStyles = makeStyles(
  theme => ({
    title: {
      marginRight: -5,
      marginBottom: theme.spacing(2)
    },
    labelValue: {
      marginBottom: theme.spacing(3),
      '&:last-of-type': { marginBottom: 0 }
    },
    section: { marginBottom: theme.spacing(6) }
  }),
  { name: 'SuperCampaignOverviewDetail' }
)

function SuperCampaignOverviewDetail() {
  const classes = useStyles()
  const superCampaign = useSuperCampaign()

  const { isLoading, mutateAsync } = useUpdateSuperCampaign()
  const [editDrawerParam, setEditDrawerParam, deleteEditDrawerParam] =
    useReplaceQueryParam('edit-drawer')

  const isReadOnly = isSuperCampaignReadOnly(superCampaign)

  const openDrawer = () => setEditDrawerParam('open')

  const closeDrawer = () => deleteEditDrawerParam()

  const handleTemplateChange = async (template: IMarketingTemplateInstance) => {
    await mutateAsync({
      superCampaign,
      inputData: { template_instance: template }
    })
  }

  const handleSuperCampaignConfirm = async (
    formValues: SuperCampaignFormValues
  ) => {
    await mutateAsync({ superCampaign, inputData: formValues })
    closeDrawer()
  }

  return (
    <SuperCampaignCard>
      <div className={classes.section} data-test="super-campaign-details">
        <SuperCampaignCardHeader
          className={classes.title}
          title="Details"
          actionLabel={!isReadOnly ? 'Edit' : undefined}
          onActionClick={openDrawer}
        />
        <SuperCampaignOverviewDetailLabelValue
          className={classes.labelValue}
          label="Description"
          value={superCampaign.description || '-'}
        />
      </div>
      {superCampaign.template_instance && (
        <SuperCampaignTemplate
          titleClassName={classes.title}
          titleVariant="body1"
          template={superCampaign.template_instance}
          onTemplateChange={handleTemplateChange}
          readOnly={isReadOnly}
          viewAsAdmin
        />
      )}
      <SuperCampaignDrawer
        isOpen={!!editDrawerParam && !isReadOnly}
        onClose={closeDrawer}
        formInitialValues={superCampaign}
        onConfirm={handleSuperCampaignConfirm}
        actionButtonsDisabled={isLoading}
      />
    </SuperCampaignCard>
  )
}

export default SuperCampaignOverviewDetail
