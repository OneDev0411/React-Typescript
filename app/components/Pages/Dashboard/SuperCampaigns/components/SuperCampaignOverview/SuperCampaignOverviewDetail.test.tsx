import * as t from '@testing-library/react'
import userEvent from '@testing-library/user-event'

import templateInstance from 'fixtures/marketing-center/template-instance.json'
import { TestBed } from 'tests/unit/TestBed'
import { makeControlledAsync } from 'tests/unit/utils/controllable-promise'

import { SuperCampaignProvider } from '../SuperCampaignProvider'

import SuperCampaignOverviewDetail from './SuperCampaignOverviewDetail'

const mockUpdateCampaignModel = makeControlledAsync(jest.fn())

jest.mock(
  '@app/models/super-campaign/update-super-campaign/update-super-campaign',
  () => async (superCampaignId: UUID, data: ISuperCampaignInput) => {
    await mockUpdateCampaignModel.fn(superCampaignId, data)

    return data
  }
)

const mockDateInMilliSeconds = new Date('2021-11-17T21:30:00').getTime()

const empty: ISuperCampaign<'template_instance'> = {
  id: '1',
  created_at: 1,
  updated_at: 1,
  type: 'super_campaign',

  brand: '123',
  created_by: '456',
  tags: null,
  eligible_brands: ['123'],
  // @ts-ignore: importing fixtures from JSON upsets TS
  template_instance: templateInstance
}
const testData: Record<string, ISuperCampaign<'template_instance'>> = {
  empty,
  filled: {
    ...empty,
    subject: 'Market Report Q3',
    description:
      'This is a monthly newsletter that is sent out to Newsletter tag',
    due_at: mockDateInMilliSeconds / 1000
  }
}

interface Props {
  superCampaign: ISuperCampaign<'template_instance'>
  setSuperCampaign: (superCampaign: ISuperCampaign<'template_instance'>) => void
}

function Test({ superCampaign, setSuperCampaign }: Props) {
  return (
    <TestBed>
      <SuperCampaignProvider superCampaign={superCampaign}>
        <SuperCampaignOverviewDetail />
      </SuperCampaignProvider>
    </TestBed>
  )
}

function testRenderEmpty() {
  const $ = t.render(
    <Test superCampaign={testData.empty} setSuperCampaign={jest.fn()} />
  )

  expect($.getByTestId('super-campaign-details')).toMatchSnapshot(
    'details box without template'
  )
  expect($.getByText('Edit')).toBeInTheDocument()
}

function testRenderFilledWithTemplate() {
  const $ = t.render(
    <Test superCampaign={testData.filled} setSuperCampaign={jest.fn()} />
  )

  expect($.getByTestId('super-campaign-details')).toMatchSnapshot(
    'details box without template'
  )
  expect(
    $.getByTestId('super-campaign-detail-template').querySelector('iframe')
  ).toMatchSnapshot('details box template preview')

  // Just checking the button exists
  t.within($.getByTestId('super-campaign-detail-template')).getByText(
    'Edit Template'
  )
}

async function testEditFlow() {
  const setSuperCampaign = jest.fn()
  const $ = t.render(
    <Test superCampaign={testData.empty} setSuperCampaign={setSuperCampaign} />
  )

  // Drawer is not yet open
  expect(t.screen.getByText('Enter Campaign Details')).not.toBeVisible()

  // Clicking on Edit will open the drawer
  userEvent.click($.getByText('Edit'))
  expect(t.screen.getByText('Enter Campaign Details')).toBeVisible()

  // Fill in the drawer form
  const form = $.getByTestId('super-campaign-edit-form')
  const find = (testId: string, selector: string) => {
    const res = t.within(form).getByTestId(testId).querySelector(selector)

    if (!res) {
      throw new Error(`Element ${selector} not found.`)
    }

    return res
  }

  userEvent.type(
    find('subject', 'input'),
    `{selectall}${testData.filled.subject}`
  )
  userEvent.type(
    find('description', 'textarea'),
    `{selectall}${testData.filled.description}`
  )

  // Click on Save
  const saveButton = $.getByText('Save')

  t.fireEvent.click(saveButton)

  // Expect `updateSuperCampaign` model to have been called correctly
  mockUpdateCampaignModel.resolve()
  await t.waitFor(() =>
    expect(mockUpdateCampaignModel.mockFn).toHaveBeenCalledTimes(1)
  )

  const expected = expect.objectContaining({
    subject: testData.filled.subject,
    description: testData.filled.description
  })

  expect(mockUpdateCampaignModel.mockFn).toHaveBeenCalledWith(
    testData.empty.id,
    expected
  )

  // Expect `Provider.setSuperCampaign` to have been called correctly
  expect(setSuperCampaign).toHaveBeenCalledWith(expected)
}

function testReadOnlyMode() {
  const data = {
    ...testData.filled,
    executed_at: Date.now() / 1000
  }
  const $ = t.render(<Test superCampaign={data} setSuperCampaign={jest.fn()} />)

  expect($.queryByText('Edit')).toBeNull()
}

describe('Super Campaign/Overview/Detail', () => {
  beforeAll(() => {
    jest
      .spyOn(Date.prototype, 'getTime')
      .mockImplementation(() => mockDateInMilliSeconds)
  })

  afterAll(() => {
    jest.restoreAllMocks()
  })

  describe('When not executed', () => {
    it('renders empty when draft campaign is given', testRenderEmpty)
    it('renders template when there is one', testRenderFilledWithTemplate)
    it.skip('completes the edit super campaign flow', testEditFlow)
  })
  describe('When executed', () => {
    it('renders in read-only mode', testReadOnlyMode)
  })
})
