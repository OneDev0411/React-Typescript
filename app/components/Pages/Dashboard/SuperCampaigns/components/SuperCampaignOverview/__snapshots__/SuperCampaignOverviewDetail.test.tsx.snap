// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Super Campaign/Overview/Detail When not executed renders empty when draft campaign is given: details box without template 1`] = `
<div
  class="SuperCampaignOverviewDetail-section-3"
  data-test="super-campaign-details"
>
  <div
    class="MuiBox-root MuiBox-root-8 SuperCampaignOverviewDetail-title-1"
  >
    <p
      class="MuiTypography-root SuperCampaignCardHeader-title-6 MuiTypography-body1"
    >
      Details
    </p>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-text SuperCampaignCardHeader-action-7 MuiButton-textPrimary MuiButton-textSizeSmall MuiButton-sizeSmall"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiButton-label"
      >
        Edit
      </span>
    </button>
  </div>
  <div
    class="SuperCampaignOverviewDetail-labelValue-2"
  >
    <span
      class="MuiTypography-root SuperCampaignOverviewDetailLabelValue-label-9 MuiTypography-caption"
    >
      Description
    </span>
    <p
      class="MuiTypography-root SuperCampaignOverviewDetailLabelValue-value-10 MuiTypography-body2"
    >
      -
    </p>
  </div>
</div>
`;

exports[`Super Campaign/Overview/Detail When not executed renders template when there is one: details box template preview 1`] = `
<iframe
  class="MarketingTemplateThumbnail-iframe-52"
  frameborder="0"
  scrolling="no"
  srcdoc="<html><body><div>Hello, World!</body></html>"
  title="Preview"
  width="700px"
/>
`;

exports[`Super Campaign/Overview/Detail When not executed renders template when there is one: details box without template 1`] = `
<div
  class="SuperCampaignOverviewDetail-section-40"
  data-test="super-campaign-details"
>
  <div
    class="MuiBox-root MuiBox-root-45 SuperCampaignOverviewDetail-title-38"
  >
    <p
      class="MuiTypography-root SuperCampaignCardHeader-title-43 MuiTypography-body1"
    >
      Details
    </p>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-text SuperCampaignCardHeader-action-44 MuiButton-textPrimary MuiButton-textSizeSmall MuiButton-sizeSmall"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiButton-label"
      >
        Edit
      </span>
    </button>
  </div>
  <div
    class="SuperCampaignOverviewDetail-labelValue-39"
  >
    <span
      class="MuiTypography-root SuperCampaignOverviewDetailLabelValue-label-46 MuiTypography-caption"
    >
      Description
    </span>
    <p
      class="MuiTypography-root SuperCampaignOverviewDetailLabelValue-value-47 MuiTypography-body2"
    >
      This is a monthly newsletter that is sent out to Newsletter tag
    </p>
  </div>
</div>
`;
